pkg/build/controller/controller.go

```diff
@@ -102,9 +102,13 @@ func (bc *BuildController) HandleBuild(build *buildapi.Build) error {
 	}
 
 	if buildutil.IsBuildComplete(build) {
+		if _, ok := build.Annotations[buildapi.BuildCompletedAnnotation]; ok {
+			return nil
+		}
```

Comment:
How is this safe?  Why aren't you just creating a cache of "really complete" builds in memory and consulting that?